TestBenches/VMRouterCM_test.cpp

```diff
@@ -0,0 +1,158 @@
+  static VMStubMEMemoryCM<outputType, rzSizeME, phiRegSize> memoryME;
+  static VMStubTEOuterMemoryCM<outputType,rzSizeTE,phiRegSize,numTEOCopies> memoryTEO;
+
+
+  ///////////////////////////
+  // Loop over events
+
+  cout << "Start event loop ..." << endl;  
+
+  // Error count
+  int err = 0;
+
+  for (unsigned int ievt = 0; ievt < nEvents; ++ievt) {
+    cout << "Event: " << dec << ievt << endl;
+
+    // Clear output memories
+    for (int i=0; i<nASCopies; ++i) {
+      memoriesAS[i].clear();
+    }
+    for (int i=0; i<nASInnerCopies; ++i) {
+      memoriesASInner[i].clear();
+    }
+    memoryME.clear();
+    if (nVMSTE) {
+      memoryTEO.clear();
+    }
+    auto path = tb.fileNames(inputPattern);
+    for (auto i = path.begin(); i != path.end(); ++i)
+    std::cout << *i << ' ';
+    
+    // Read event and write to memories
+    for (unsigned int i = 0; i < numInputs; i++) {
+      writeMemFromFile<InputStubMemory<inputType>>(inputStubs[i], fin_inputstubs[i], ievt);
+    }
+    for (unsigned int i = 0; i < numInputsDisk2S; i++) {
+      writeMemFromFile<InputStubMemory<DISK2S>>(inputStubsDisk2S[i], fin_inputstubs_disk2s[i], ievt);
+    }
+
+    // bx - bunch crossing
+    BXType bx = ievt;
+    BXType bx_out;
+
+    // Unit Under Test
+    VMRouterCMTop(bx, bx_out, inputStubs
+#if kDISK > 0
+        , inputStubsDisk2S
+#endif
+        , memoriesAS
+#if kLAYER == 1 || kLAYER == 2 || kLAYER == 3 || kLAYER == 5 || kDISK == 1 || kDISK == 3
+        , memoriesASInner
+#endif
+        , &memoryME
+#if kLAYER == 2 || kLAYER == 3 || kLAYER == 4 || kLAYER == 6 || kDISK == 1 || kDISK == 2 || kDISK == 4
+        , &memoryTEO
+#endif
+      );
+
+    // Compare the computed outputs with the expected ones
+    // Add 1 to the error count per stub that is incorrect
+    bool truncation = false;
+
+    // AllStub memories
+    for (unsigned int i = 0; i < numASCopies; i++) {
+      err += compareMemWithFile<AllStubMemory<outputType>>(memoriesAS[i], fout_allstubs[i], ievt, "AllStub", truncation);
+    }
+    // Allstub Inner memories
+    if (nASInnerCopies) {
+      for (unsigned int i = 0; i < numASInnerCopies; i++) {
+        err += compareMemWithFile<AllStubInnerMemory<outputType>>(memoriesASInner[i], fout_allstubs_inner[i], ievt, "AllStubInner", truncation);
+      }
+    }
+    // ME memories
+    err += compareBinnedMemWithFile<VMStubMEMemoryCM<outputType, rzSizeME, phiRegSize>>(memoryME, fout_vmstubme[0], ievt, "VMStubME", truncation);  
+    //TE Outer memories
+    if (nVMSTE) {
+      err += compareBinnedMemCMWithFile<VMStubTEOuterMemoryCM<outputType, rzSizeTE, phiRegSize, numTEOCopies>>(memoryTEO, fout_vmstubte[0], ievt, "VMStubTEOuter", truncation);
+    }
+  } // End of event loop
+
+	cerr << "Exiting with return value " << err << endl;
+	// This is necessary because HLS seems to only return an 8-bit error count, so if err%256==0, the test bench can falsely pass
+	if (err > 255) err = 255;
+	return err;
```

Comment:
Please move these lines to the left so they are inline with the bracket at the end of the event loop.